### ISSUE TITLE
Update operations do not support joins

### ISSUE THREAD

**Issue by kibertoad** [tgriesser/knex#2796]
# Environment

Knex version: 0.15.2

# Bug

```
  await knex('table)
    .innerJoin('table2', function() {
      this.on(
        `table.id`,
        `table2.id`
      ).andOn(`table.field`, 'value')
    })
    .update({
      table.field2: 'newValue,
    })
```

This is going to execute without an error, but entire join part will be ignored and all entries will be silently updated. 

https://github.com/tgriesser/knex/issues/1534 suggests either using raw or subquery, but neither of these solutions is particularly elegant, and API-wise there is no reason why this shouldn't work, so I would consider this a bug.

**Comment by kibertoad** [comment]
I could try giving it a shot, but I have no idea how complicated this is going to be. @elhigu any pointers where to start?

**Comment by elhigu** [comment]
Not bug, but a missing feature. This has been discussed a lot during history and is one of the most wanted features in my list. I don't know if it would be hard to implement or not. I would suggest reading builder / compilation code of update operation to see if join parameters are even collected to builder for update operations.

**Comment by elhigu** [comment]
IIRC postgresql also supports kind of joins in updates, but it uses from keyword to implement them. I'm not sure how much different they are from joins. Mssql also seems to support it. For oracle there doesn't seem to be anything else but hacks for it https://stackoverflow.com/a/2446859/360060

**Comment by woodgates** [comment]
This is needed as update with joins is a very common operation.

**Comment by JQuezada0** [comment]
May I take a stab at this?

**Comment by kibertoad** [comment]
@JQuezada0 Yes, that would be great! You might want to take a look at how DELETE with joins was implemented:
https://github.com/knex/knex/pull/4568
https://github.com/knex/knex/pull/4800
https://github.com/knex/knex/pull/5016

**Comment by OlivierCavadenti** [comment]
@JQuezada0 you can ping me here for any questions if you want.

**Comment by jpike88** [comment]
+1 for this, I've switched from mysql to Postgres, and this was a shock... considering that Postgres actually doing everything SQL compliant and others aren't.

**Comment by jpike88** [comment]
The worst part is that the query fails silently, if using Postgres it should at least throw an error.

**Comment by jpike88** [comment]
I think it's worth implementing a stopgap solution. Perhaps if it was possible to add some sort of validation method on update that can inspect the AST of the query before it's executed.

**Comment by jpike88** [comment]
Another one that is a gotcha for PostgresSQL users... this knex command will silently create an invalid query:


```
await sql
			.knexMaster('user')
			.join(sql.Table.Business, 'user.businessID', 'business.id')
			.where('user.id', creatorID)
			.increment('bookingCounter', 1);
```

The need for dialect-specific checks is definitely there.

**Comment by kibertoad** [comment]
Pull requests are very welcome.

**Comment by jpike88** [comment]
Which would be more ideal:

- The ability to import a Typescript definition set that overrides existing types for a given dialect (a type that excludes join when update or increment is present), I'm not sure that's even possible
- Throw errors at runtime

**Comment by joelanman** [comment]
Could this at least throw an error or warning for now - that this isn't supported? I'm using Postgres and the error was hard to track down to this issue: `missing FROM-clause entry for table`

**Comment by kibertoad** [comment]
@joelanman PR would be appreciated!

**Comment by mercmobily** [comment]
This is an update on this long standing issue.

Support for UPDATE with JOIN depends on dialect today, and this inconsistency is the core problem. MySQL and SQL Server compilers include JOINs in UPDATE, but PostgreSQL and SQLite compilers do not consider JOINs for UPDATE at all, which leads to the confusing behaviour you observed on those dialects.

Some code reference to help future coding: the MySQL compiler explicitly includes JOINs in UPDATE, see `lib/dialects/mysql/query/mysql-querycompiler.js:107` where `const join = this.join();` is included and then used in `update()` at `lib/dialects/mysql/query/mysql-querycompiler.js:116`. SQL Server similarly compiles UPDATE with `from ... join`, see `lib/dialects/mssql/query/mssql-querycompiler.js:170` onward where `join` is used in `updateWithTriggers()` at `lib/dialects/mssql/query/mssql-querycompiler.js:187`.

By contrast, PostgreSQL and SQLite update compilers do not use the builder joins for UPDATE. PostgreSQL’s `update()` compiles without joins and only supports an explicit `from` via `.updateFrom()`, see `lib/dialects/postgres/query/pg-querycompiler.js:49` and `_updateFrom()` at `lib/dialects/postgres/query/pg-querycompiler.js:145`. SQLite’s `update()` similarly omits any JOIN handling, see `lib/dialects/sqlite3/query/sqlite-querycompiler.js:153`.

Minimal reproductions using toSQL (no database needed):

```js
// SQLite3: join is ignored in UPDATE
const kSqlite = require('./')({ client: 'sqlite3', useNullAsDefault: true });
console.log(
  kSqlite('t').join('u', 't.id', 'u.id').update({ x: 1 }).toSQL().sql
);
// Output: update `t` set `x` = ?

// PostgreSQL: join is ignored in UPDATE; use updateFrom instead
const kPg = require('./')({ client: 'pg' });
console.log(
  kPg('t').join('u', 't.id', 'u.id').update({ x: 1 }).toSQL().sql
);
// Output: update "t" set "x" = ?
console.log(
  kPg('t').update({ x: 1 }).updateFrom('u').whereRaw('t.id = u.id').toSQL().sql
);
// Output: update "t" set "x" = ? from "u" where t.id = u.id

// MySQL: join is included in UPDATE
const kMy = require('./')({ client: 'mysql' });
console.log(
  kMy('t').join('u', 't.id', 'u.id').update({ x: 1 }).toSQL().sql
);
// Output: update `t` inner join `u` on `t`.`id` = `u`.`id` set `x` = ?
```

PostgreSQL exposes `.updateFrom(tableName)` for this exact case, documented at `docs/src/guide/query-builder.md:887` with example usage at `docs/src/guide/query-builder.md:894`.

The issue remains open as a feature request to unify behaviour and/or validation. It would be beneficial to either a) add support per dialect where possible and expose a cross-dialect API, or b) detect and throw when a JOIN is present on UPDATE for dialects that don’t support it, to prevent silent surprises.

Current workarounds:

- PostgreSQL: use `.updateFrom()` with a WHERE clause to express the join condition.
- All dialects: use a subquery filter, for example:

```js
await knex('table')
  .update({ field2: 'newValue' })
  .whereIn('id',
    knex('table')
      .select('table.id')
      .innerJoin('table2', 'table.id', 'table2.id')
      .where('table.field', 'value')
  );
```

- Raw SQL for dialect-specific syntax, if preferred.

Related items for context and precedent:
- PR #2946 (closed, not merged) – Support for joins in update statements
- PR #4568 (closed) – implement delete with join
- PR #4800 (closed) – Support Joins and Using syntax for PostgreSQL Delete Statement (#4591)
- PR #5016 (closed) – Fix delete joins with PostgreSQL
